platform/pc/service.c

```diff
@@ -470,8 +471,8 @@ void detect_hypervisor(kernel_heaps kh)
         if (!xen_detect(kh)) {
             if (!hyperv_detect(kh)) {
                 init_debug("no hypervisor detected; assuming qemu full emulation");
-                if (!init_hpet(kh)) {
-                    halt("HPET initialization failed; no timer source\n");
+                if (!init_hpet(kh) && !init_tsc_timer(kh)) {
```

Comment:
As Justin noted, it may be preferable to use a calibrated TSC and default to HPET as a last resort. And, for the sake of debugging, could this report which timer was initialized via init_debug?

Reply:
OK, now the calibrated TSC takes precedence over the HPET, and INIT_DEBUG shows which timer is being used.